File: search_matrix.py
def search_matrix():
    matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]]
    target = 60

    for r in matrix:
        if r[-1] < target:
            continue
        else:
            l, h = 0, len(matrix[0])
            while l <= h:
                m = l + ((h - l) // 2)
                if r[m] > target:
                    h = m - 1
                elif r[m] < target:
                    l = m + 1
                else:
                    return True

    return False

print(search_matrix())

File: search_matrix.rs
use std::cmp::Ordering::{Equal, Less, Greater};

fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
    for row in &matrix {
        if let Some(&last_element) = row.last() {
            if last_element < target {
                continue;
            } else {
                let (mut l, mut r) = (0, row.len());
                while l < r {
                    let m = l + (r - l) / 2;
                    match target.cmp(&row[m]) {
                        Equal => return true,
                        Less => r = m,
                        Greater => l = m + 1,
                    }
                }
            }
        }
    }
    false
}

fn main() {
    let matrix: Vec<Vec<i32>> = vec![
        vec![1, 3, 5, 7],
        vec![10, 11, 16, 20],
        vec![23, 30, 34, 60],
    ];
    let target = -1;
    println!("{:?}", search_matrix(matrix, target));
}
